vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,666 @@
+// the output_example function has special logic to ensure the number of feaures is correctly calculated.
+// Copy anything in default namespace for slot to ccb_slot_namespace in shared
+// Copy other slot namespaces to shared
+void inject_slot_features(example* shared, example* slot)
+{
+  for (auto index : slot->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+    else if (index == default_namespace)  // slot default namespace has a special namespace in shared
+    {
+      LabelDict::add_example_namespace(*shared, ccb_slot_namespace, slot->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::add_example_namespace(*shared, index, slot->feature_space[index]);
+    }
+  }
+}
+
+template <bool audit>
+void inject_slot_id(ccb& data, example* shared, int id)
+{
+  // id is zero based, so the vector must be of size id + 1
+  if(id + 1 > data.hashes.size())
+  {
+    data.hashes.resize(id + 1, 0);
+  }
+
+  uint64_t index;
+  if(data.hashes[id] == 0)
+  {
+    auto current_index_str = "index"+std::to_string(id);
+    index = VW::hash_feature(*data.all, current_index_str, data.id_namespace_hash);
+    data.hashes[id] = index;
+  }
+  else
+  {
+    index = data.hashes[id];
+  }
+
+  shared->feature_space[ccb_id_namespace].push_back(1., index);
+  shared->indices.push_back(ccb_id_namespace);
+
+  if (audit)
+  {
+    auto current_index_str = "index"+std::to_string(id);
+    shared->feature_space[ccb_id_namespace].space_names.push_back(audit_strings_ptr(new audit_strings(data.id_namespace_str, current_index_str)));
+  }
+}
+
+// Since the slot id is the only thing in this namespace, the popping the value off will work correctly.
+template<bool audit>
+void remove_slot_id(example* shared)
+{
+  shared->feature_space[ccb_id_namespace].indicies.pop();
+  shared->feature_space[ccb_id_namespace].values.pop();
+  shared->indices.pop();
+
+  if (audit)
+  {
+    shared->feature_space[ccb_id_namespace].space_names.pop();
+  }
+}
+
+void remove_slot_features(example* shared, example* slot)
+{
+  for (auto index : slot->indices)
+  {
+    // constant namespace should be ignored, as it already exists and we don't want to double it up.
+    if (index == constant_namespace)
+    {
+      continue;
+    }
+
+    else if (index == default_namespace)  // slot default namespace has a special namespace in shared
+    {
+      LabelDict::del_example_namespace(*shared, ccb_slot_namespace, slot->feature_space[32]);
+    }
+    else
+    {
+      LabelDict::del_example_namespace(*shared, index, slot->feature_space[index]);
+    }
+  }
+}
+
+// Generates quadratics between each namespace and the slot id as well as appends slot id to every existing interaction.
+void calculate_and_insert_interactions(example* shared, std::vector<example*> actions, std::vector<std::string>& vec)
```

Comment:
There are a number of strings created and destroyed in this function.   v_array<string> w/ a modified resize() of v_array where string destructors are not called + only string assignment operations which cause a memcpy into existing string storage would save those allocations.

Reply:
I do not think this is worth it given the scale that this function operates at. I also recently moved to a set, so the number of strings that are copied into the final generated_interactions is further reduced.

Reply:
a) 'set' is generally a bad data structure because it involves lots of pointer chasing.
b) When -q :: is used, the number of strings is quite high and running through them is the dominant computational factor.

Reply:
In this function we want to maintain a unique collection of namespaces to use, which a set lends itself to well. Unless we determine this to be a significant perf bottleneck then I think a set should be used.

Reply:
I expect this is a significant perf bottleneck when -q :: is active.  

But, I'm slightly confused about 'unique collection of namespaces to use'.  Why/when would the set of namespaces not be unique by default?

Reply:
They would not be unique because each action contains some collection of namespaces and there is almost certainly overlap between them

Reply:
Different actions would certainly commonly have the same namespace interactions.  Why do you care whether they are unique?

Reply:
The vector of interactions must be unique - we need to ensure it remains as a unique set. We can either use a set as we go or once we are done remove duplicates

Reply:
1) It need not be unique---nothing breaks if it's not unique, you just spend more computation.  In particular, sometimes people explicitly choose to have (say) -q ua and -q au for redundancy reasons.
2) Given that it starts unique for individual example (=action and/or shared example) how do you lose uniqueness?

Reply:
1. In my interpretation those are still unique, UA and UA would be duplicates and not unique. Wouldn't it do too many updates for the interaction feature? Regardless we don't want to do wasted work there.
2. Uniqueness is lost because several examples may have the same namespace (very likely) and since we need to scan over every action to find the namespaces we may end up with duplicates.

I did a performance test using google benchmark. I compared using a set as I am now with string copies, and using the vector directly and using binary search to find whether to insert every interaction into the set. Interestingly, the results showed almost exactly equal performance (several runs averages out the difference). 

```
2019-06-26 13:42:50
Running ./perf
Run on (40 X 2195 MHz CPU s)
Load Average: 0.52, 0.58, 0.59
----------------------------------------------------------------
Benchmark                      Time             CPU   Iterations
----------------------------------------------------------------
BM_Set                     64458 ns        64174 ns        11200
BM_VectorBinarySearch      54268 ns        54688 ns        10000
```

Reply:
I'm still not understanding why you expect to get duplicates.  

Assuming the input interactions are unique and not using the private namespace for CCB, can you get duplicates?  How?

W.r.t. benchmark, what's the computational time when you do no duplicate detection?

Reply:
The duplicate comes from appending slot_id_namespace with action namespaces. Each action has namespace "Action" so after scanning all n actions there are n A + slot_id_namespace  interactions.

Doing no deduping:
```
2019-06-26 14:08:54
Running ./perf
Run on (40 X 2195 MHz CPU s)
Load Average: 0.52, 0.58, 0.59
----------------------------------------------------------------
Benchmark                      Time             CPU   Iterations
----------------------------------------------------------------
BM_Set                      5184 ns         5162 ns       112000
BM_VectorBinarySearch       5132 ns         5156 ns       100000
BM_NoUnique                 2582 ns         2623 ns       280000
```

Reply:
So this is definitely a 'hot' spot.  

If the interactions for every action are the same, then why do we need to do everything n times?  Can we just skip that and never have dups?

Reply:
We don't know what interactions are present for any example until we see it. 

Perhaps we have a misunderstanding here. Should I instead be simply inserting a single interaction in the form {':','slot_id_namespace'}? Currently I am scanning for what's present and calculating the interactions?

Reply:
I agree that we don't know interactions in advance, but we do know that all the interactions of any individual example are unique if they should be.  

You can't use ':' because that needs to be expanded at this level. 

I'm still fundamentally confused about why we should care about duplicates.  It seems like we should be able to process each action independently without ever suffering through a union.

Reply:
I was reviewing GPerf output.  It turns out that Time is time / iteration.  The number of iterations is not relevant except that you don't want something like one iteration.  So BM_VectorBinarySearch is better.

Reply:
Ah I see, that makes sense. Thanks for the clarification!

We're using a constant sized vector to mark now, and we don't need to dedupe so we are in a better place.